crates/autopilot/src/shadow.rs

```diff
@@ -0,0 +1,270 @@
+                return None;
+            }
+        };
+
+        if self.auction == auction.id {
+            tracing::trace!("skipping already seen auction");
+            return None;
+        }
+        if self.block == auction.auction.block {
+            tracing::trace!("skipping already seen block");
+            return None;
+        }
+
+        if auction
+            .auction
+            .orders
+            .iter()
+            .all(|order| match order.metadata.class {
+                OrderClass::Market => false,
+                OrderClass::Liquidity => true,
+                OrderClass::Limit(_) => false,
+            })
+        {
+            tracing::trace!("skipping empty auction");
+            return None;
+        }
+
+        self.auction = auction.id;
+        self.block = auction.auction.block;
+        Some(auction)
+    }
+
+    async fn single_run(&self, id: AuctionId, auction: Auction) {
+        tracing::info!("solving");
+        Metrics::get().auction.set(id);
+
+        let mut participants = self.competition(id, &auction).await;
+
+        // Shuffle so that sorting randomly splits ties.
+        participants.shuffle(&mut rand::thread_rng());
+        participants.sort_unstable_by_key(|participant| cmp::Reverse(participant.score()));
+
+        if let Some(Participant {
+            driver,
+            solution: Ok(solution),
+        }) = participants.first()
+        {
+            let reference_score = participants
+                .get(1)
+                .map(|participant| participant.score())
+                .unwrap_or_default();
+            let reward = solution
+                .score
+                .get()
+                .checked_sub(reference_score)
+                .expect("reference score unexpectedly larger than winner's score");
+
+            tracing::info!(
+                driver =% driver.url,
+                score =% solution.score,
+                %reward,
+                "winner"
+            );
+            Metrics::get()
+                .performance_rewards
+                .with_label_values(&[driver.url.as_str()])
+                .inc_by(reward.to_f64_lossy());
+        }
+
+        let hex = |bytes: &[u8]| format!("0x{}", hex::encode(bytes));
+        for Participant { driver, solution } in participants {
+            match solution {
+                Ok(solution) => {
+                    tracing::debug!(
+                        driver =% driver.url,
+                        score =% solution.score,
+                        account =? solution.account,
+                        calldata =% hex(&solution.calldata.internalized),
+                        "participant"
+                    );
+                    if solution.calldata.internalized != solution.calldata.uninternalized {
+                        tracing::debug!(
+                            driver =% driver.url,
+                            uninternalized =% hex(&solution.calldata.uninternalized),
```

Comment:
Would be nice if this could be folded into the other log.

Reply:
c567f420c2b438976e12ace04484258a670c7577

Reply:
> Also the log says `internalized interactions` when it prints the `uninternalized` data.

This was intentional - the log was indicating that the solver chose to internalize interactions, and then prints the full uninternalized calldata in the message. Makes sense to me at least 😅.

Reply:
Ahh, I see. Seemed a bit unintuitive to me but I can see where you are coming from now.